emacs/.emacs.d/config.org
#+title: GNU Emacs Config
#+author: Petar Đurašković
#+description: Petar Đurašković personal GNU Emacs configuration.
#+startup: fold
#+options: toc:2
#+auto_tangle: t
#+PROPERTY: header-args :tangle "./config.el"

* Table of contents :toc:
- [[#important-things-to-load-first][Important things to load first]]
  - [[#env][Env]]
  - [[#straight-package-manager][Straight Package Manager]]
  - [[#theme][Theme]]
  - [[#transient][Transient]]
- [[#browser][Browser]]
  - [[#bookmarks][Bookmarks]]
  - [[#elpher-gemini][Elpher (gemini)]]
- [[#custom-functions][Custom Functions]]
  - [[#cooking-timer][Cooking timer]]
  - [[#yank-n-times][Yank N times]]
- [[#consult][Consult]]
- [[#dired][Dired]]
- [[#emms][EMMS]]
- [[#email-mu4e][Email (mu4e)]]
- [[#exwm][EXWM]]
- [[#graphical-user-interface-tweaks][Graphical User Interface tweaks]]
  - [[#disable-menubar-toolbars-and-scrollbars][Disable Menubar, Toolbars and Scrollbars]]
  - [[#disable-font-lock-mode][Disable font-lock-mode]]
  - [[#display-line-numbers-and-truncated-lines][Display line numbers and truncated lines]]
  - [[#read-extended-command-predicate][Read extended command predicate]]
  - [[#setting-the-font-face][Setting The Font Face]]
- [[#sly][SLY]]
- [[#programming-things][Programming things]]
  - [[#corfu][Corfu]]
  - [[#general-compiling-settings][General compiling settings]]
  - [[#electric-pair-mod][Electric pair mod]]
  - [[#rust-settings][Rust settings]]
  - [[#prog-mode][Prog mode]]
- [[#sudo][Sudo]]
- [[#magit][Magit]]
- [[#multiple-cursors][Multiple cursors]]
- [[#modes][Modes]]
  - [[#repeat-mode][Repeat mode]]
- [[#orderless][Orderless]]
- [[#org-mode][Org Mode]]
  - [[#auto-tangle-org-file][Auto tangle org file]]
  - [[#enabling-table-of-contents][Enabling Table of Contents]]
  - [[#enabling-org-bullets][Enabling Org bullets]]
  - [[#disable-electric-indent][Disable electric indent]]
  - [[#enable-org-tempo][Enable org tempo]]
  - [[#disable-code-block-highlighting][Disable code block highlighting]]
  - [[#setting-header-font-size][Setting header font size]]
  - [[#hooks][Hooks]]
- [[#rss-elfeed][RSS (elfeed)]]
- [[#pacmacs][Pacmacs]]
- [[#pdf-tools][Pdf-tools]]
- [[#vertico][Vertico]]
- [[#vterm][Vterm]]
- [[#which-key][Which key]]
- [[#wttrin][Wttrin]]
- [[#keybindings][Keybindings]]
  - [[#ibuffer][Ibuffer]]
  - [[#config][Config]]
  - [[#compiling][Compiling]]
  - [[#elfeed][Elfeed]]
  - [[#emms-1][Emms]]
  - [[#exwm-1][EXWM]]
  - [[#yank-n-times-1][Yank N times]]
  - [[#vterm-1][Vterm]]

* Important things to load first
Packages and configurations are not in alphabetical order! If they
where things wouldn't work properly.

** Env
#+begin_src emacs-lisp
(setenv "LC_COLLATE" "C")
(setq x-hyper-keysym 'Hyper_L)

(put 'upcase-region 'disabled nil)
(put 'downcase-region 'disabled nil)
(setq url-user-agent "curl")
#+end_src

** Straight Package Manager
#+begin_src emacs-lisp
(defvar bootstrap-version)
(let ((bootstrap-file
       (expand-file-name
        "straight/repos/straight.el/bootstrap.el"
        (or (bound-and-true-p straight-base-dir)
            user-emacs-directory)))
      (bootstrap-version 7))
  (unless (file-exists-p bootstrap-file)
    (with-current-buffer
        (url-retrieve-synchronously
         "https://raw.githubusercontent.com/radian-software/straight.el/develop/install.el"
         'silent 'inhibit-cookies)
      (goto-char (point-max))
      (eval-print-last-sexp)))
  (load bootstrap-file nil 'nomessage))

(straight-use-package 'use-package)
(straight-use-package 'org)
#+end_src

** Theme
#+begin_src emacs-lisp
(load-theme 'tango-dark t)
#+end_src
Mu4e and magit are complaining about transient so just load it from straight at the beginning
and forget about it.
** Transient
#+begin_src emacs-lisp
(use-package transient
  :straight t)
#+end_src
* Browser
** Bookmarks
#+begin_src emacs-lisp
(setq bookmark-save-flag 1)
#+end_src

** Elpher (gemini)
#+begin_src emacs-lisp
  (use-package elpher
    :straight (elpher :type git :host github :repo "emacsmirror/elpher")
    :bind ("C-c g" . elpher)
    :config
    (setq elpher-open-urls-with-eww t)
    (setq elpher-user-header t))
#+end_src

* Custom Functions
** Cooking timer
#+begin_src emacs-lisp
  (defvar zeko/cooking-timer-process nil "Timer process for the cooking timer.")
  (defvar zeko/cooking-timer-remaining-time -1 "Remaining time in seconds for cooking timer.")
  (defvar zeko/cooking-timer-sound "~/.emacs.d/sounds/sabaton.mp3" "Sound that's being played when timer is finished.")

  (defun zeko/cooking-timer-countdown ()
    "Function that updates echo area with remaining time."
    (setq zeko/cooking-timer-remaining-time (1- zeko/cooking-timer-remaining-time))
    (let ((minutes (floor zeko/cooking-timer-remaining-time 60))
          (seconds (% zeko/cooking-timer-remaining-time 60))
  	      (message-log-max nil)) ;; Prevent message logging into *Messages* buffer
      (message "Timer: %02d:%02d left" minutes seconds)
      (when (eq zeko/cooking-timer-remaining-time 0)
        (message "Timer: Done!")
        (emms-play-file zeko/cooking-timer-sound)
        (cancel-timer zeko/cooking-timer-process)
        (setq zeko/cooking-timer-process nil)
        (setq zeko/cooking-timer-remaining-time -1))))

  (defun zeko/cooking-timer-cancel ()
    "Cancel the running cooking timer."
    (interactive)
    (if zeko/cooking-timer-process
        (progn
          (cancel-timer zeko/cooking-timer-process)
          (setq zeko/cooking-timer-process nil)
          (setq zeko/cooking-timer-remaining-time -1)
          (message "Timer canceled..."))
      (message "No timer running...")))

  (defun zeko/cooking-timer-start ()
    (interactive)
    "Play a sound after period of time in minutes"
    (let ((number (read-number "Time in minutes: ")))
      (if zeko/cooking-timer-process
          (message "Timer is already running, cancel it first!")
        (progn
          (setq zeko/cooking-timer-remaining-time (* number 60))
          (setq zeko/cooking-timer-process
                (run-at-time 0 1 #'zeko/cooking-timer-countdown))))))
#+end_src

** Yank N times
#+begin_src emacs-lisp
(defun zeko/yank-n-times (n)
  "Yank (paste) the top of the kill buffer N times."
  (interactive "p")
  (dotimes (_ n)
    (yank)
    (newline)))
#+end_src
* Consult
#+begin_src emacs-lisp
  (use-package consult
    :straight t
    :bind (
           ("C-c s r" . consult-ripgrep)
  	     ("C-c s f" . consult-fd)
  	     ("H-l"     . consult-line))
    :custom
    (consult-preview-key 'any) 
    (consult-async-min-input 2))
#+end_src

* Dired
#+begin_src emacs-lisp
  (use-package dired-preview
    :straight t
    :config
    ;; set limit to 12 MB
    (setq dired-preview-max-size (* 12 1024 1024)))
  (add-hook 'dired-mode-hook (lambda () (dired-preview-mode 1)))
  (setq dired-listing-switches "-lha --time-style=+%d/%m/%Y")
  (setf dired-kill-when-opening-new-dired-buffer t)
#+end_src

* EMMS
#+begin_src emacs-lisp
  (use-package emms
    :straight t
    :commands (zeko/emms-playlist-new-named emms-smart-browse emms-start)
    :init
    (defun zeko/capitalize-words (str)
      "Capitalize each word in the input string STR, and make the rest of the string lowercase."
      (mapconcat
       (lambda (word)
         (concat (upcase (substring word 0 1))
                 (downcase (substring word 1))))
       (split-string str)
       " "))

    :config
    (require 'emms-setup)
    (require 'emms-playlist-mode)
    (emms-standard)
    (emms-default-players)
    
    (setq emms-cache-file "~/.emacs.d/emms/cache")
    (setq emms-info-asynchronously t)

    (defun zeko/emms-playlist-new-named (name)
      "Create and switch to new EMMS playlist named NAME."
      (interactive "sEnter playlist name: ")
      (let* ((capitalized-name (zeko/capitalize-words name))
             (formatted-name (format "*EMMS: %s*" capitalized-name))
             (buf (emms-playlist-new formatted-name))
             (dir (expand-file-name (concat "~/Music/" capitalized-name))))

        (unless (file-directory-p dir)
          (kill-buffer buf)
          (cl-return-from zeko/emms-playlist-new-named (format "Directory `%s` not found!" dir)))

        (switch-to-buffer buf)
        (emms-playlist-set-playlist-buffer buf)
        (emms-add-directory dir))))
#+end_src

* Email (mu4e)
#+begin_src emacs-lisp
  (setq user-full-name "Petar Đurašković")
  (setq user-mail-address "developmentpetar@gmail.com")

  (use-package mu4e
    :ensure nil
    ;; As on Arch you have to have load-path
    :load-path "/usr/share/emacs/site-lisp/mu4e/"
    :config
    (setq mu4e-change-filenames-when-moving t)
    (setq mu4e-get-mail-command "mbsync -a")
    (setq mu4e-maildir "~/.emacs.d/Mail")
    (setq mu4e-update-interval (* 60 60))

    (setq mu4e-drafts-folder  "/[Gmail]/Drafts")
    (setq mu4e-sent-folder    "/[Gmail]/Sent Mail")
    (setq mu4e-refile-folder  "/[Gmail]/All Mail")
    (setq mu4e-trash-folder   "/[Gmail]/Trash")

    (require 'auth-source-pass)
    (auth-source-pass-enable)
    (setq auth-sources '(password-store))

    (setq message-send-mail-function 'smtpmail-send-it
  	    smtpmail-smtp-user "developmentpetar@gmail.com"
  	    smtpmail-smtp-server "smtp.gmail.com"
          smtpmail-smtp-service 587
          smtpmail-stream-type 'starttls)

    (setq mu4e-maildir-shortcuts
          '(("/Inbox"                 . ?i)
            ("/[Gmail]/Sent Mail"     . ?s)
            ("/[Gmail]/Trash"         . ?t)
            ("/[Gmail]/Drafts Mail"   . ?d)
            ("/[Gmail]/Starred"       . ?f)
            ("/[Gmail]/All Mail"      . ?a))))

  (keymap-set global-map "C-c e" #'mu4e)
#+end_src

* EXWM
#+begin_src emacs-lisp
  (use-package exwm
    :straight t
    :if (string= (getenv "CURRENT_WM") "exwm")
    :config
    (setq exwm-workspace-number 4)

    (add-hook 'exwm-update-class-hook
              (lambda () (exwm-workspace-rename-buffer exwm-class-name)))

    (setq exwm-input-global-keys
          `(([?\s-r] . exwm-reset) ;; s-r: Reset (to line-mode).
            ([?\s-w] . exwm-workspace-switch) ;; s-w: Switch workspace.
            ([?\s-&] . (lambda (cmd) ;; s-&: Launch application.
            	       (interactive (list (read-shell-command "$ ")))
            	       (start-process-shell-command cmd nil cmd)))

    	      ([?\s-1] . (lambda () (interactive) (exwm-workspace-switch-create 0)))
            ([?\s-2] . (lambda () (interactive) (exwm-workspace-switch-create 1)))
            ([?\s-3] . (lambda () (interactive) (exwm-workspace-switch-create 2)))
            ([?\s-4] . (lambda () (interactive) (exwm-workspace-switch-create 3)))

  	      ([?\s-!] . (lambda () (interactive) (exwm-workspace-move-window 0)))
            ([?\s-@] . (lambda () (interactive) (exwm-workspace-move-window 1)))
            ([?\s-#] . (lambda () (interactive) (exwm-workspace-move-window 2)))
            ([?\s-$] . (lambda () (interactive) (exwm-workspace-move-window 3)))))

    (setq exwm-randr-workspace-monitor-plist
      	'(0 "eDP-1"
      	  1 "eDP-1"
      	  2 "HDMI-1"
      	  3 "HDMI-1"))
    (add-hook 'exwm-randr-screen-change-hook
              (lambda ()
                (start-process-shell-command
                 "xrandr" nil "xrandr --output HDMI-1 --above eDP-1 --mode 1920x1080")))
    (exwm-randr-mode 1)
    (display-battery-mode 1)

    (setq display-time-day-and-date t)
    (setq display-time-24hr-format t)
    (display-time-mode 1)

    (exwm-enable))
#+end_src

* Graphical User Interface tweaks
Make GNU Emacs look a bit better...

** Disable Menubar, Toolbars and Scrollbars
#+begin_src emacs-lisp
  (menu-bar-mode -1)
  (tool-bar-mode -1)
  (scroll-bar-mode -1)
#+end_src

** Disable font-lock-mode
#+begin_src emacs-lisp
  (defvar zeko/font-lock-whitelist
    '(org-mode
      dired-mode
      ibuffer-mode
      vterm-mode
      wttrin-mode)
    "List of modes that have font-lock-mode enabled.")

  (defun zeko/around-font-lock-mode (orig-fun &rest args)
    "Prevent font-lock from turning on unless the mode is whitelisted."
    (if (memq major-mode zeko/font-lock-whitelist)
        (apply orig-fun args)
      (apply orig-fun '(-1))))

  ;; No confetti
  (advice-add 'font-lock-mode :around #'zeko/around-font-lock-mode)
#+end_src

** Display line numbers and truncated lines
#+begin_src emacs-lisp
(setq display-line-numbers-type 'relative)
(global-visual-line-mode t)
(add-hook 'text-mode-hook #'display-line-numbers-mode)
#+end_src

** Read extended command predicate

emacs/.emacs.d/config.el
(setenv "LC_COLLATE" "C")
(setq x-hyper-keysym 'Hyper_L)

(put 'upcase-region 'disabled nil)
(put 'downcase-region 'disabled nil)
(setq url-user-agent "curl")

(defvar bootstrap-version)
(let ((bootstrap-file
       (expand-file-name
        "straight/repos/straight.el/bootstrap.el"
        (or (bound-and-true-p straight-base-dir)
            user-emacs-directory)))
      (bootstrap-version 7))
  (unless (file-exists-p bootstrap-file)
    (with-current-buffer
        (url-retrieve-synchronously
         "https://raw.githubusercontent.com/radian-software/straight.el/develop/install.el"
         'silent 'inhibit-cookies)
      (goto-char (point-max))
      (eval-print-last-sexp)))
  (load bootstrap-file nil 'nomessage))

(straight-use-package 'use-package)
(straight-use-package 'org)

(load-theme 'tango-dark t)

(use-package transient
  :straight t)

(setq bookmark-save-flag 1)

(use-package elpher
  :straight (elpher :type git :host github :repo "emacsmirror/elpher")
  :bind ("C-c g" . elpher)
  :config
  (setq elpher-open-urls-with-eww t)
  (setq elpher-user-header t))

(defvar zeko/cooking-timer-process nil "Timer process for the cooking timer.")
(defvar zeko/cooking-timer-remaining-time -1 "Remaining time in seconds for cooking timer.")
(defvar zeko/cooking-timer-sound "~/.emacs.d/sounds/sabaton.mp3" "Sound that's being played when timer is finished.")

(defun zeko/cooking-timer-countdown ()
  "Function that updates echo area with remaining time."
  (setq zeko/cooking-timer-remaining-time (1- zeko/cooking-timer-remaining-time))
  (let ((minutes (floor zeko/cooking-timer-remaining-time 60))
        (seconds (% zeko/cooking-timer-remaining-time 60))
	      (message-log-max nil)) ;; Prevent message logging into *Messages* buffer
    (message "Timer: %02d:%02d left" minutes seconds)
    (when (eq zeko/cooking-timer-remaining-time 0)
      (message "Timer: Done!")
      (emms-play-file zeko/cooking-timer-sound)
      (cancel-timer zeko/cooking-timer-process)
      (setq zeko/cooking-timer-process nil)
      (setq zeko/cooking-timer-remaining-time -1))))

(defun zeko/cooking-timer-cancel ()
  "Cancel the running cooking timer."
  (interactive)
  (if zeko/cooking-timer-process
      (progn
        (cancel-timer zeko/cooking-timer-process)
        (setq zeko/cooking-timer-process nil)
        (setq zeko/cooking-timer-remaining-time -1)
        (message "Timer canceled..."))
    (message "No timer running...")))

(defun zeko/cooking-timer-start ()
  (interactive)
  "Play a sound after period of time in minutes"
  (let ((number (read-number "Time in minutes: ")))
    (if zeko/cooking-timer-process
        (message "Timer is already running, cancel it first!")
      (progn
        (setq zeko/cooking-timer-remaining-time (* number 60))
        (setq zeko/cooking-timer-process
              (run-at-time 0 1 #'zeko/cooking-timer-countdown))))))

(defun zeko/yank-n-times (n)
  "Yank (paste) the top of the kill buffer N times."
  (interactive "p")
  (dotimes (_ n)
    (yank)
    (newline)))

(use-package consult
  :straight t
  :bind (
         ("C-c s r" . consult-ripgrep)
	     ("C-c s f" . consult-fd)
	     ("H-l"     . consult-line))
  :custom
  (consult-preview-key 'any) 
  (consult-async-min-input 2))

(use-package dired-preview
  :straight t
  :config
  ;; set limit to 12 MB
  (setq dired-preview-max-size (* 12 1024 1024)))
(add-hook 'dired-mode-hook (lambda () (dired-preview-mode 1)))
(setq dired-listing-switches "-lha --time-style=+%d/%m/%Y")
(setf dired-kill-when-opening-new-dired-buffer t)

(use-package emms
  :straight t
  :commands (zeko/emms-playlist-new-named emms-smart-browse emms-start)
  :init
  (defun zeko/capitalize-words (str)
    "Capitalize each word in the input string STR, and make the rest of the string lowercase."
    (mapconcat
     (lambda (word)
       (concat (upcase (substring word 0 1))
               (downcase (substring word 1))))
     (split-string str)
     " "))

  :config
  (require 'emms-setup)
  (require 'emms-playlist-mode)
  (emms-standard)
  (emms-default-players)
  
  (setq emms-cache-file "~/.emacs.d/emms/cache")
  (setq emms-info-asynchronously t)

  (defun zeko/emms-playlist-new-named (name)
    "Create and switch to new EMMS playlist named NAME."
    (interactive "sEnter playlist name: ")
    (let* ((capitalized-name (zeko/capitalize-words name))
           (formatted-name (format "*EMMS: %s*" capitalized-name))
           (buf (emms-playlist-new formatted-name))
           (dir (expand-file-name (concat "~/Music/" capitalized-name))))

      (unless (file-directory-p dir)
        (kill-buffer buf)
        (cl-return-from zeko/emms-playlist-new-named (format "Directory `%s` not found!" dir)))

      (switch-to-buffer buf)
      (emms-playlist-set-playlist-buffer buf)
      (emms-add-directory dir))))

(setq user-full-name "Petar Đurašković")
(setq user-mail-address "developmentpetar@gmail.com")

(use-package mu4e
  :ensure nil
  ;; As on Arch you have to have load-path
  :load-path "/usr/share/emacs/site-lisp/mu4e/"
  :config
  (setq mu4e-change-filenames-when-moving t)
  (setq mu4e-get-mail-command "mbsync -a")
  (setq mu4e-maildir "~/.emacs.d/Mail")
  (setq mu4e-update-interval (* 60 60))

  (setq mu4e-drafts-folder  "/[Gmail]/Drafts")
  (setq mu4e-sent-folder    "/[Gmail]/Sent Mail")
  (setq mu4e-refile-folder  "/[Gmail]/All Mail")
  (setq mu4e-trash-folder   "/[Gmail]/Trash")

  (require 'auth-source-pass)
  (auth-source-pass-enable)
  (setq auth-sources '(password-store))

  (setq message-send-mail-function 'smtpmail-send-it
	    smtpmail-smtp-user "developmentpetar@gmail.com"
	    smtpmail-smtp-server "smtp.gmail.com"
        smtpmail-smtp-service 587
        smtpmail-stream-type 'starttls)

  (setq mu4e-maildir-shortcuts
        '(("/Inbox"                 . ?i)
          ("/[Gmail]/Sent Mail"     . ?s)
          ("/[Gmail]/Trash"         . ?t)
          ("/[Gmail]/Drafts Mail"   . ?d)
          ("/[Gmail]/Starred"       . ?f)
          ("/[Gmail]/All Mail"      . ?a))))

(keymap-set global-map "C-c e" #'mu4e)

(use-package exwm
  :straight t
  :if (string= (getenv "CURRENT_WM") "exwm")
  :config
  (setq exwm-workspace-number 4)

  (add-hook 'exwm-update-class-hook
            (lambda () (exwm-workspace-rename-buffer exwm-class-name)))

  (setq exwm-input-global-keys
        `(([?\s-r] . exwm-reset) ;; s-r: Reset (to line-mode).
          ([?\s-w] . exwm-workspace-switch) ;; s-w: Switch workspace.
          ([?\s-&] . (lambda (cmd) ;; s-&: Launch application.
          	       (interactive (list (read-shell-command "$ ")))
          	       (start-process-shell-command cmd nil cmd)))

  	      ([?\s-1] . (lambda () (interactive) (exwm-workspace-switch-create 0)))
          ([?\s-2] . (lambda () (interactive) (exwm-workspace-switch-create 1)))
          ([?\s-3] . (lambda () (interactive) (exwm-workspace-switch-create 2)))
          ([?\s-4] . (lambda () (interactive) (exwm-workspace-switch-create 3)))

	      ([?\s-!] . (lambda () (interactive) (exwm-workspace-move-window 0)))
          ([?\s-@] . (lambda () (interactive) (exwm-workspace-move-window 1)))
          ([?\s-#] . (lambda () (interactive) (exwm-workspace-move-window 2)))
          ([?\s-$] . (lambda () (interactive) (exwm-workspace-move-window 3)))))

  (setq exwm-randr-workspace-monitor-plist
    	'(0 "eDP-1"
    	  1 "eDP-1"
    	  2 "HDMI-1"
    	  3 "HDMI-1"))
  (add-hook 'exwm-randr-screen-change-hook
            (lambda ()
              (start-process-shell-command
               "xrandr" nil "xrandr --output HDMI-1 --above eDP-1 --mode 1920x1080")))
  (exwm-randr-mode 1)
  (display-battery-mode 1)

  (setq display-time-day-and-date t)
  (setq display-time-24hr-format t)
  (display-time-mode 1)

  (exwm-enable))

(menu-bar-mode -1)
(tool-bar-mode -1)
(scroll-bar-mode -1)

(defvar zeko/font-lock-whitelist
  '(org-mode
    dired-mode
    ibuffer-mode
    vterm-mode
    wttrin-mode)
  "List of modes that have font-lock-mode enabled.")

(defun zeko/around-font-lock-mode (orig-fun &rest args)
  "Prevent font-lock from turning on unless the mode is whitelisted."
  (if (memq major-mode zeko/font-lock-whitelist)
      (apply orig-fun args)
    (apply orig-fun '(-1))))

;; No confetti
(advice-add 'font-lock-mode :around #'zeko/around-font-lock-mode)

(setq display-line-numbers-type 'relative)
(global-visual-line-mode t)
(add-hook 'text-mode-hook #'display-line-numbers-mode)

(setq read-extended-command-predicate
      #'command-completion-default-include-p)

(set-face-attribute 'default nil
  		    :font "JetBrains Mono"
   		    :height 120
   		    :weight 'medium)

(set-fontset-font t 'unicode
                  (font-spec :family "Noto Color Emoji"))

;; Uncomment the following line if spacing needs adjusting
(setq-default line-spacing 0.12)

;; When emacsclient starts, there was this weird issue of it not having font that was
;; specified in the configuration. The nil value in set-face-attribure is for current
;; frame. Since emacsclient doesn't have current frame as it runs at the backround as
;; a daemon, it had no way to apply font to it. This add-to-list fixes it.
(add-to-list 'default-frame-alist '(font . "JetBrains Mono-12"))

(use-package sly 
  :straight t
  :config
  (setq inferior-lisp-program "/usr/bin/sbcl"))

(use-package corfu
  :straight t
  :init
  (global-corfu-mode -1) 
  (add-hook 'emacs-lisp-mode-hook #'corfu-mode)
  (add-hook 'lisp-mode-hook #'corfu-mode)
  (add-hook 'sly-mrepl-mode-hook #'corfu-mode)
  (add-hook 'lisp-interaction-mode-hook #'corfu-mode))

(defvar zeko/mode-list
  '((rust-mode . "cargo run")
    (c-mode    . "make")
    (c++-mode  . "make"))
  "Mapping major programming modes to compile commands.")

(defun zeko/language-options (option)
  "Provide build/compile/test option and switch to that buffer."
  (when option
    (compile option)
    (select-window (get-buffer-window "*compilation*"))))

(defun zeko/compile-for-major-mode ()
  "Get the compile command for the current major mode and run it."
  (interactive)
  (if-let ((command (alist-get major-mode zeko/mode-list)))
    (progn
	    (compile command)
      ;; Small delay to get the window appear, then select it
      (select-window (get-buffer-window "*compilation*")))
    (message "No compile command defined for %s" major-mode)))

(electric-pair-mode 1)

;; Disable auto-pairing globally by default using the trick above
(setq-default electric-pair-inhibit-predicate
              (lambda (char) (not (use-region-p))))

(add-hook 'emacs-lisp-mode-hook
          (lambda () 
            (setq-local electric-pair-inhibit-predicate #'electric-pair-default-inhibit)))

(add-hook 'lisp-mode-hook
          (lambda () 
            (setq-local electric-pair-inhibit-predicate #'electric-pair-default-inhibit)))

(add-hook 'sly-mrepl-mode-hook
          (lambda () 
            (setq-local electric-pair-inhibit-predicate #'electric-pair-default-inhibit)))

(use-package rust-mode 
  :straight t
  :mode ("\\.rs\\'" . rust-mode)
  :hook (rust-mode . (lambda () (face-remap-add-relative 'default :height 140))))

(add-hook 'conf-toml-mode-hook (lambda () (display-line-numbers-mode)))

(defun zeko/prog-mode-setup ()
  "My custom settings for all programming modes."
  (display-line-numbers-mode 1))

(add-hook 'prog-mode-hook #'zeko/prog-mode-setup)

(use-package sudo-edit
  :straight t
  :bind (
	 ("H-x s f" . sudo-edit-find-file)
	 ("H-x s e" . sudo-edit)))

(use-package magit 
  :straight t 
  :after transient)

(use-package multiple-cursors
  :straight t
  :bind (
  ("C-S-c C-S-c" . mc/edit-lines)
  ("C->"         . mc/mark-next-like-this)
  ("C-<"         . mc/mark-previous-like-this)
  ("C-c C-<"     . mc/mark-all-like-this)))

(repeat-mode 1)

(use-package orderless
  :straight t
  :custom
  (completion-styles '(orderless basic))
  (completion-category-defaults nil)
  (completion-category-overrides '((file (styles partial-completion)))))

(defun zeko/org-babel-tangle-on-save ()
  "Tangle the current buffer if #+auto_tangle: t is present."
  (when (and (eq major-mode 'org-mode)
	         'buffer-modified-p
             (save-excursion
               (goto-char (point-min))
               (search-forward "#+auto_tangle: t" nil t)))
    (let ((org-confirm-babel-evaluate nil))
      (org-babel-tangle))))

(add-hook 'after-save-hook #'zeko/org-babel-tangle-on-save)

(use-package toc-org
  :straight t
  :commands toc-org-enable
  :init (add-hook 'org-mode-hook 'toc-org-enable))

(add-hook 'org-mode-hook 'org-indent-mode)
(use-package org-bullets :straight t)
(add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))

(electric-indent-mode -1)

(require 'org-tempo)

(setq org-src-fontify-natively nil)

(defun zeko/org-faces ()
    (set-face-attribute 'org-todo nil :height 0.9)
    (set-face-attribute 'org-document-title nil :height 2.0 :foreground "white")
    (set-face-attribute 'org-level-1 nil :height 1.7)
    (set-face-attribute 'org-level-2 nil :height 1.6)
    (set-face-attribute 'org-level-3 nil :height 1.5)
    (set-face-attribute 'org-level-4 nil :height 1.4)
    (set-face-attribute 'org-level-5 nil :height 1.3)
    (set-face-attribute 'org-level-6 nil :height 1.2)
    (set-face-attribute 'org-level-7 nil :height 1.1)
    (set-face-attribute 'org-level-8 nil :height 1.0))

(add-hook 'org-mode-hook (lambda () (zeko/org-faces)  			   
			   (face-remap-add-relative 'default :height 130)))

(use-package elfeed :straight t)

(use-package elfeed-org
  :straight t
  :after elfeed
  :config
  (setq rmh-elfeed-org-files '("~/org/elfeed.org"))
  (elfeed-org))
        
(defun zeko/elfeed-show-eww-if-tag (entry tag)
  "Browse elfeed ENTRY in eww if it is tagged with TAG."
  (when (member tag (elfeed-entry-tags entry))
    (let ((browse-url-browser-function #'eww-browse-url))
      (elfeed-show-visit))))

(defun zeko/elfeed-show-eww-if-tag-is-browse (entry)
  "Browse elfeed ENTRY in eww if it has the tag `browse'."
  (zeko/elfeed-show-eww-if-tag entry 'browse))

(defun zeko/elfeed-tag-buzzwords (entry)
  "Remove buzzwords and topics from elfeed."
  (let ((title (elfeed-entry-title entry)))
    (when (string-match-p
 	   (rx (or "ai" "llm" "openai" "ai-generated" "chatgpt"))
	   (downcase title))
      (elfeed-tag entry 'buzzword))))

(setq-default elfeed-search-filter "@3-months-ago -buzzword")
(advice-add #'elfeed-show-entry :after #'zeko/elfeed-show-eww-if-tag-is-browse)
(add-hook 'elfeed-new-entry-hook #'zeko/elfeed-tag-buzzwords)

(use-package pacmacs 
  :straight t
  :hook (pacmacs-mode . (lambda () 
			  (setq line-spacing 0))))

(use-package pdf-tools
  :straight t
  :mode ("\\.pdf\\'" . pdf-view-mode)
  :hook (pdf-view-mode . pdf-isearch-minor-mode))

(use-package vertico
  :straight t
  :init
  (vertico-mode))

(custom-set-faces '(vertico-current ((t (:background "#2e3436" :inherit bold)))))

(use-package vterm 
  :straight t
  :config
  (defun zeko/vterm-dont-ask-on-kill ()
    (setq-local kill-buffer-query-functions nil))

  (add-hook 'vterm-mode-hook 'zeko/vterm-dont-ask-on-kill))

(use-package which-key
  :straight t
  :init (which-key-mode 1)
  :config
  (setq which-key-popup-type 'minibuffer
    which-key-side-window-location 'bottom
    which-key-frame-max-width 6
	which-key-idle-delay 0.4
    which-key-frame-max-height 30
	which-key-max-description-length 30
	which-key-unicode-correction 3
    which-key-separator " → "))
(which-key-add-key-based-replacements "C-c l" "LSP/Eglot")

(use-package wttrin :straight t)
(keymap-set global-map "C-c w w" (lambda () (interactive) (wttrin "Podgorica")))

(defun zeko/ibuffer-toggle ()
  "Toggle ibuffer buffer."
  (interactive)
  (if-let ((buf (get-buffer "*Ibuffer*")))
      (switch-to-buffer buf)
    (ibuffer)))

(keymap-set global-map "<f9>" #'zeko/ibuffer-toggle)

(with-eval-after-load 'ibuffer
  (define-prefix-command 'zeko/ibuffer)
  (keymap-set global-map "C-c i" 'zeko/ibuffer)

  (keymap-set 'zeko/ibuffer "g" #'ibuffer-toggle-filter-group)
  (keymap-set 'zeko/ibuffer "r" #'ibuffer-filter-disable)
  (keymap-set 'zeko/ibuffer "a" #'ibuffer-switch-to-saved-filter-groups))

(defun zeko/open-config ()
  "Open my main Emacs config."
  (interactive)
  (find-file (expand-file-name "~/.emacs.d/config.org")))

(defun zeko/reload-init-file ()
  "Reload Emacs init file."
  (interactive)
  (load-file user-init-file))

(define-prefix-command 'zeko/config)
(keymap-set global-map "C-c c" 'zeko/config)

(keymap-set 'zeko/config "e" #'zeko/open-config)
(keymap-set 'zeko/config "r" #'zeko/reload-init-file)

(keymap-set global-map "H-c" #'zeko/compile-for-major-mode)

;; Rust-specific "Run Options"
(with-eval-after-load 'rust-mode
  (keymap-set rust-mode-map "H-x r" (lambda () (interactive) (zeko/language-options "cargo build --release")))
  (keymap-set rust-mode-map "H-x t" (lambda () (interactive) (zeko/language-options "cargo test"))))

(define-prefix-command 'zeko/elfeed)
(keymap-set global-map "C-c f" 'zeko/elfeed)

(keymap-set 'zeko/elfeed "o" #'elfeed)
(keymap-set 'zeko/elfeed "u" #'elfeed-update)

(define-prefix-command 'zeko/emms)
(keymap-set global-map "C-c m" 'zeko/emms)

(keymap-set 'zeko/emms "d" #'zeko/emms-playlist-new-named)
(keymap-set 'zeko/emms "n" #'emms-next)
(keymap-set 'zeko/emms "m" #'emms-shuffle)
(keymap-set 'zeko/emms "l" #'emms-previous)
(keymap-set 'zeko/emms "q" #'emms-stop)
(keymap-set 'zeko/emms "p" #'emms-pause)
(keymap-set 'zeko/emms "f" #'emms-play-file)
(keymap-set 'zeko/emms "k" #'keep-lines)

(defvar-keymap zeko-shuffle-repeat-map
  :doc "Repeat map for shuffling"
  :repeat t
  "m" #'emms-shuffle)

(keymap-set global-map "<f1>" (lambda () (interactive) (shell-command "pactl set-sink-mute @DEFAULT_SINK@ toggle")))
(keymap-set global-map "<f2>" (lambda () (interactive) (shell-command "pactl set-sink-volume @DEFAULT_SINK@ -5%")))
(keymap-set global-map "<f3>" (lambda () (interactive) (shell-command "pactl set-sink-volume @DEFAULT_SINK@ +5%")))

(keymap-set global-map "<f5>" (lambda () (interactive) (shell-command "brightnessctl set 5%-")))
(keymap-set global-map "<f6>" (lambda () (interactive) (shell-command "brightnessctl set +5%")))

(keymap-set global-map "H-y" #'zeko/yank-n-times)

(define-prefix-command 'zeko/vterm)
(keymap-set zeko/vterm "c" #'vterm-send-C-c)

(keymap-set global-map "C-c t" #'vterm)

(with-eval-after-load 'vterm
  (keymap-set vterm-mode-map "C-c t" 'zeko/vterm))
